libraries/AP_Scripting/examples/plane_ship_landing.lua

```diff
@@ -0,0 +1,439 @@
+   local y_sq = sq(wind:y())
+   local closing_speed
+   if tas_sq >= y_sq then
+      closing_speed = math.sqrt(tas_sq - y_sq)
+   else
+      -- min 1 m/s
+      closing_speed = 1.0
+   end
+
+   -- include the wind in the direction of the ship
+   closing_speed = closing_speed + wind:x()
+
+   -- account for the ship velocity
+   closing_speed = closing_speed - ship2d:x()
+
+   -- calculate stopping distance
+   return sq(closing_speed) / (2.0 * Q_TRANS_DECEL:get())
+end
+
+-- get holdoff distance
+function get_holdoff_distance()
+   local radius = WP_LOITER_RAD:get()
+   local holdoff_dist = math.abs(radius*1.5)
+   local stop_distance = stopping_distance()
+
+   -- increase holdoff distance by up to 50% to ensure we can stop
+   holdoff_dist = math.max(holdoff_dist, math.min(holdoff_dist*2.5, stop_distance*2))
+   return holdoff_dist
+end
+
+-- get the holdoff position
+function get_holdoff_position()
+   local radius = WP_LOITER_RAD:get()
+   local heading_deg = target_heading + SHIP_LAND_ANGLE:get()
+   local holdoff_dist = get_holdoff_distance()
+
+   local ofs = Vector2f()
+   ofs:x(-holdoff_dist)
+   ofs:y(radius)
+   ofs:rotate(math.rad(heading_deg))
+   local target = target_pos:copy()
+   target:offset(ofs:x(), ofs:y())
+   return target
+end
+
+function wrap_360(angle)
+   local res = math.fmod(angle, 360.0)
+    if res < 0 then
+        res = res + 360.0
+    end
+    return res
+end
+
+function wrap_180(angle)
+    local res = wrap_360(angle)
+    if res > 180 then
+       res = res - 360
+    end
+    return res
+end
+
+--[[
+   check if we have reached the tangent to the landing location
+--]]
+function check_approach_tangent()
+   local distance = current_pos:get_distance(target_pos)
+   local holdoff_dist = get_holdoff_distance()
+   if landing_stage == STAGE_HOLDOFF and throttle_pos <= THROTTLE_MID and distance < 4*holdoff_dist then
+      gcs:send_text(0, string.format("Descending for approach (hd=%.1fm)", holdoff_dist))
+      landing_stage = STAGE_DESCEND
+   end
+   if reached_alt and landing_stage == STAGE_DESCEND then
+      -- go to approach stage when throttle is low, we are
+      -- pointing at the ship and have reached target alt.
+      -- Also require we are within 2.5 radius of the ship, and our heading is within 20
+      -- degrees of the target heading
+      local target_bearing_deg = wrap_180(math.deg(current_pos:get_bearing(target_pos)))
+      local ground_bearing_deg = wrap_180(math.deg(ahrs:groundspeed_vector():angle()))
+      local margin = 10
+      local distance = current_pos:get_distance(target_pos)
+      local holdoff_dist = get_holdoff_distance()
+      local error1 = math.abs(wrap_180(target_bearing_deg - ground_bearing_deg))
+      local error2 = math.abs(wrap_180(ground_bearing_deg - (target_heading + SHIP_LAND_ANGLE:get())))
+      logger.write('SLND','TBrg,GBrg,Dist,HDist,Err1,Err2','ffffff',target_bearing_deg, ground_bearing_deg, distance, holdoff_dist, error1, error2)
```

Comment:
you can add units and multipliers if your felling keen. eg 

`logger.write('SLND','TBrg,GBrg,Dist,HDist,Err1,Err2','ffffff','ddmmd','------',target_bearing_deg`...

Reply:
i'm happy without the extra info for this debug msg